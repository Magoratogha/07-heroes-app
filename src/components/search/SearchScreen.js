import { useNavigate, useLocation } from "react-router-dom";
import queryString from 'query-string';
import getHeroesByName from "../../helpers/getHeroesByName";
import useForm from "../../hooks/useForm";
import HeroCard from "../hero/HeroCard";
import { useMemo } from "react";

function SearchScreen() {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [ formValue, handleValueChange ] = useForm({
        search: q,
    });

    const { search } = formValue;
    const filteredHeroes = useMemo(() => getHeroesByName(q), [q]);
    
    const handleSubmit = (e) => {
        e.preventDefault();
        navigate(`?q=${search}`);
    }

    return (
        <div className="row">
            <div className="col-5">
                <h4>Search</h4>
                <hr></hr>
                <form onSubmit={ handleSubmit }>
                    <input
                        onChange={ handleValueChange }
                        value={ search }
                        type="text"
                        className="form-control"
                        placeholder="Type something..."
                        name="search"
                        autoComplete="off"></input>
                    <button
                        className="btn btn-outline-dark mt-2">
                            Submit
                    </button>
                </form>
            </div>
            <div className="col-7">
                <h4>Results</h4>
                <hr></hr>
                {
                    q === '' ?
                        <div className="alert alert-primary">
                            Type something in the search box to search for a hero.
                        </div>
                    :
                        filteredHeroes.length > 0 ?
                            filteredHeroes.map(hero => (
                                <HeroCard key={ hero.id } { ...hero } />
                            ))
                        :
                            <div className="alert alert-danger">
                                No results found.
                            </div>
                }
            </div>
        </div>
    );
}

export default SearchScreen;