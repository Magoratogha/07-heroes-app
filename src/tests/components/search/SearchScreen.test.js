import { queryByAttribute, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter, useNavigate } from "react-router-dom";
import SearchScreen from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockNavigate,
}));

describe('Test on <SearchScreen />', () => { 

    const user = userEvent.setup()
    
    it('should render correctly', () => {
        const { container } = render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>);
        expect(container).toMatchSnapshot();
        expect(screen.getByText('Type something in the search box to search for a hero.', { selector: 'div' }))
            .toBeInTheDocument();
    });

    it('should render batman card and fill the input with the queryString', () => {
        const { container } = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>);
        expect(screen.getByText('Batman', { selector: 'h5' })).toBeInTheDocument();
        expect(queryByAttribute('class', container, 'form-control')).toHaveValue('batman');
    });

    it('should render alert when the queryString is invalid', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=123']}>
                <SearchScreen />
            </MemoryRouter>);
        expect(screen.getByText('No results found.', { selector: 'div' })).toBeInTheDocument();
    });

    it('should call handleSubmit when the form is submitted', async() => {
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>);
        const input = screen.getByPlaceholderText('Type something...', { exact: true });
        await user.type(input, 'batman');
        await user.keyboard('{enter}');
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
    });
});