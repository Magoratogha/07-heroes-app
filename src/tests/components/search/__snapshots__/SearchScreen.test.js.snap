// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <SearchScreen /> should render correctly 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="search"
          placeholder="Type something..."
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-dark mt-2"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary"
      >
        Type something in the search box to search for a hero.
      </div>
    </div>
  </div>
</div>
`;
